src/test/run-pass/async-fn-size.rs

```diff
@@ -0,0 +1,106 @@
+// edition:2018
+// aux-build:arc_wake.rs
+
+#![feature(async_await, await_macro)]
+
+extern crate arc_wake;
+
+use std::pin::Pin;
+use std::future::Future;
+use std::sync::{
+    Arc,
+    atomic::{self, AtomicUsize},
+};
+use std::task::{Context, Poll};
+use arc_wake::ArcWake;
+
+struct Counter {
+    wakes: AtomicUsize,
+}
+
+impl ArcWake for Counter {
+    fn wake(self: Arc<Self>) {
+        Self::wake_by_ref(&self)
+    }
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        arc_self.wakes.fetch_add(1, atomic::Ordering::SeqCst);
+    }
+}
+
+struct WakeOnceThenComplete(bool, u8);
+
+impl Future for WakeOnceThenComplete {
+    type Output = u8;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<u8> {
+        if self.0 {
+            Poll::Ready(self.1)
+        } else {
+            cx.waker().wake_by_ref();
+            self.0 = true;
+            Poll::Pending
+        }
+    }
+}
+
+fn wait(fut: impl Future<Output = u8>) -> u8 {
+    let mut fut = Box::pin(fut);
+    let counter = Arc::new(Counter { wakes: AtomicUsize::new(0) });
+    let waker = ArcWake::into_waker(counter.clone());
+    let mut cx = Context::from_waker(&waker);
+    loop {
+        match fut.as_mut().poll(&mut cx) {
+            Poll::Ready(out) => return out,
+            Poll::Pending => (),
+        }
+    }
+}
+
+fn base() -> WakeOnceThenComplete { WakeOnceThenComplete(false, 1) }
+
+async fn await1_level1() -> u8 {
+    await!(base())
+}
+
+async fn await2_level1() -> u8 {
+    await!(base()) + await!(base())
+}
+
+async fn await3_level1() -> u8 {
+    await!(base()) + await!(base()) + await!(base())
+}
+
+async fn await3_level2() -> u8 {
+    await!(await3_level1()) + await!(await3_level1()) + await!(await3_level1())
+}
+
+async fn await3_level3() -> u8 {
+    await!(await3_level2()) + await!(await3_level2()) + await!(await3_level2())
+}
+
+async fn await3_level4() -> u8 {
+    await!(await3_level3()) + await!(await3_level3()) + await!(await3_level3())
+}
+
+async fn await3_level5() -> u8 {
+    await!(await3_level4()) + await!(await3_level4()) + await!(await3_level4())
+}
+
+fn main() {
```

Comment:
Size comparison:

```
New    Old
  2      2  base()
 12      8  await1_level1()
 12     12  await2_level1()
 12     12  await3_level1()
 24     44  await3_level2()
 36    140  await3_level3()
 48    428  await3_level4()
 60   1292  await3_level5()
```

This does increase size of `await1_level1()` by 4 bytes, because of the "prefix" trick I'm using in the layout computation. We can do better with a greedy placement algorithm, but representing the state of that algorithm is quite tricky. This prevents future sizes from multiplying each time you add a new level of `await` nesting, and I think that's much more important right now.

Reply:
Nevermind, removing the `Option` in `GeneratorLayout.storage_conflicts` made the optimization smarter, and the size of the future returned by `await1_level1()` is now the same as before!

Also, my change in #60840 to make `MaybeStorageLive` drop-aware further decreased the size of the bigger futures.

New size comparison:

```
New    Old
  2      2  base()
  8      8  await1_level1()
 12     12  await2_level1()
 12     12  await3_level1()
 20     44  await3_level2()
 28    140  await3_level3()
 36    428  await3_level4()
 44   1292  await3_level5()
```